### PR TITLE
Device context in main thread

### PR DESCRIPTION
**Pull request by CedricGuillemet**
thread init and run split in 2 functions.
It's not possible to hook a MetalLayer in a thread other than the main thread.

### CODE REVIEW

File: Library/Source/NativeEngine.cpp [hunk 1]
```diff
@@ -275,23 +275,24 @@ namespace Babylon
             bgfx::TextureFormat::RGBA32F};
     }
 
-    Napi::FunctionReference NativeEngine::InitializeAndCreateConstructor(Napi::Env& env)
+    void NativeEngine::InitializeDeviceContext(void *nativeWindowPtr, uint32_t width, uint32_t height)
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]:
```suggestion
    void NativeEngine::InitializeDeviceContext(void* nativeWindowPtr, uint32_t width, uint32_t height)
```

File: Library/Source/NativeEngine.cpp [hunk 2]
```diff
@@ -275,23 +275,24 @@ namespace Babylon
...
+        init.resolution.width = width;
+        init.resolution.height = height;
         init.resolution.reset = BGFX_RESET_FLAGS;
         bgfx::init(init);
         bgfx::setViewClear(0, BGFX_CLEAR_COLOR | BGFX_CLEAR_DEPTH, 0x443355FF, 1.0f, 0);
         bgfx::setViewRect(0, 0, 0, init.resolution.width, init.resolution.height);
         bgfx::touch(0);
+    }
 
+    Napi::FunctionReference NativeEngine::InitializeAndCreateConstructor(Napi::Env& env)
```
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: I would rename to just this:
```suggestion
    Napi::FunctionReference NativeEngine::CreateConstructor(Napi::Env& env)
```

File: Library/Source/RuntimeUWP.cpp [hunk 3]
```diff
@@ -15,6 +15,7 @@ namespace Babylon
         : Runtime{std::make_unique<RuntimeImpl>(window, rootUrl)}
         , m_window{window}
     {
+        NativeEngine::InitializeDeviceContext(window, 32, 32);
```
Comment (bghgary) [inline Library/Source/RuntimeUWP.cpp]: I know this is what it was doing before, but ideally, we really should figure out the size of the native window here and possibly all the platforms, like what Win32 one does.

File: Library/Source/RuntimeWin32.cpp [hunk 4]
```diff
@@ -18,9 +18,10 @@ namespace Babylon
         RECT rect;
         if (GetWindowRect(hWnd, &rect))
         {
-            float width = static_cast<float>(rect.right - rect.left);
-            float height = static_cast<float>(rect.bottom - rect.top);
-            UpdateSize(width, height);
+            auto width = rect.right - rect.left;
+            auto height = rect.bottom - rect.top;
+            NativeEngine::InitializeDeviceContext(hWnd, width, height);
+            UpdateSize(static_cast<float>(width), static_cast<float>(height));
```
Comment (bghgary) [inline Library/Source/RuntimeWin32.cpp]: This call is extraneous.

File: Library/Source/NativeEngine.cpp (hunk 1 above)
Comment (bghgary) [inline Library/Source/NativeEngine.cpp]: The name DeviceContext is a bit overloaded, especially with DirectX. Maybe just `Initialize`?

Comment (CedricGuillemet) [reply Library/Source/NativeEngine.cpp]: InitializeContext?
Initialize is a too vague

Comment (bghgary) [reply Library/Source/NativeEngine.cpp]: InitializeWindow?

Comment (bghgary) [reply Library/Source/NativeEngine.cpp]: Though I don't think Initialize is very vague. We are initializing bgfx inside.

**Pull request merged**